cmd/oras/internal/option/format.go

```diff
@@ -15,21 +15,127 @@ limitations under the License.
 
 package option
 
-import "github.com/spf13/pflag"
+import (
+	"fmt"
+	"strings"
 
-// Format is a flag to format metadata into output.
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	oerrors "oras.land/oras/cmd/oras/internal/errors"
+)
+
+const (
+	// format types
+	FormatTypeJSON       = "json"
+	FormatTypeTree       = "tree"
+	FormatTypeTable      = "table"
+	FormatTypeGoTemplate = "go-template"
+)
+
+// FormatType represents a custom description in help doc.
+type FormatType struct {
+	Name  string
+	Usage string
+}
+
+// Format contains input and parsed options for formatted output flags.
 type Format struct {
+	Type     string
 	Template string
+	// FormatFlag can be private once deprecated `--output` is removed from
+	// `oras discover`
+	FormatFlag string
+	types      []FormatType
 }
 
 // ApplyFlag implements FlagProvider.ApplyFlag.
 func (opts *Format) ApplyFlags(fs *pflag.FlagSet) {
-	const name = "format"
-	if fs.Lookup(name) != nil {
-		// allow command to overwrite the flag
-		return
-	}
-	fs.StringVar(&opts.Template, name, "", `[Experimental] Format output using a custom template:
-'json':       Print in JSON format
-'$TEMPLATE':  Print output using the given Go template.`)
+	usage := "[Experimental] Format output using a custom template:"
+	if len(opts.types) == 0 {
+		opts.types = []FormatType{
+			{Name: FormatTypeJSON, Usage: "Print in JSON format"},
+			{Name: FormatTypeGoTemplate, Usage: "Print output using the given Go template"},
+		}
+	}
+
+	// generate usage string
+	maxLength := 0
+	for _, option := range opts.types {
+		if len(option.Name) > maxLength {
+			maxLength = len(option.Name)
+		}
+	}
+	for _, option := range opts.types {
+		usage += fmt.Sprintf("\n'%s':%s%s", option.Name, strings.Repeat(" ", maxLength-len(option.Name)+2), option.Usage)
+	}
+
+	// apply flags
+	fs.StringVar(&opts.FormatFlag, "format", opts.FormatFlag, usage)
+	fs.StringVar(&opts.Template, "template", "", "[Experimental] Template string used to format output")
+}
+
+// Parse parses the input format flag.
+func (opts *Format) Parse(_ *cobra.Command) error {
+	if err := opts.parseFlag(); err != nil {
+		return err
+	}
+
+	if opts.Type == "" {
+		// flag not specified
+		return nil
+	}
+
+	var optionalTypes []string
+	for _, option := range opts.types {
+		if opts.Type == option.Name {
+			// type validation passed
+			return nil
+		}
+		optionalTypes = append(optionalTypes, option.Name)
+	}
+	return &oerrors.Error{
+		Err:            fmt.Errorf("invalid format type: %q", opts.Type),
+		Recommendation: fmt.Sprintf("supported types: %s", strings.Join(optionalTypes, ", ")),
+	}
+}
+
+func (opts *Format) parseFlag() error {
+	opts.Type = opts.FormatFlag
+	if opts.Template != "" {
+		// template explicitly set
+		if opts.Type != FormatTypeGoTemplate {
+			return fmt.Errorf("--template must be used with --format %s", FormatTypeGoTemplate)
```

Comment:
There is a case that we may also consider add error catch and error message:

If I use `--format go-template` only (without `--template`), the default output and verbose log are empty. See examples below.

It would be reasonable to print an error message `Error: you should use "--format go-template=GO_TEMPLATE"` to prompt users that this is a wrong usage and how to use this flag.

```
$ oras manifest fetch localhost:5000/my-ubuntu:latest --format go-template

$ oras manifest fetch localhost:5000/my-ubuntu:latest --format go-template -v
```

Reply:
Will align the error reporting to `kubectl`
```console
> kubectl get all --output go-template
error: template format specified but no template given
> kubectl get all --output go-template=""
error: template format specified but no template given
```